src/libcore/iter/adapters/mod.rs

```diff
@@ -1609,6 +1609,100 @@ impl<I: Iterator, P> Iterator for TakeWhile<I, P>
     }
 }
 
+/// An iterator that only accepts elements while `predicate` returns `Some(_)`.
+///
+/// This `struct` is created by the [`take_while_map`] method on [`Iterator`]. See its
+/// documentation for more.
+///
+/// [`take_while_map`]: trait.Iterator.html#method.take_while
+/// [`Iterator`]: trait.Iterator.html
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+#[unstable(feature = "iter_take_while_map", reason = "recently added", issue = "0")]
+#[derive(Clone)]
+pub struct TakeWhileMap<I, P> {
+    iter: I,
+    flag: bool,
+    predicate: P,
+}
+
+impl<I, P> TakeWhileMap<I, P> {
+    pub(super) fn new(iter: I, predicate: P) -> TakeWhileMap<I, P> {
+        TakeWhileMap { iter, flag: false, predicate }
+    }
+}
+
+#[unstable(feature = "iter_take_while_map", reason = "recently added", issue = "0")]
+impl<I: fmt::Debug, P> fmt::Debug for TakeWhileMap<I, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("TakeWhileMap")
+            .field("iter", &self.iter)
+            .field("flag", &self.flag)
+            .finish()
+    }
+}
+
+#[unstable(feature = "iter_take_while_map", reason = "recently added", issue = "0")]
+impl<B, I: Iterator, P> Iterator for TakeWhileMap<I, P>
+where P: FnMut(I::Item) -> Option<B>
+{
+    type Item = B;
+
+    #[inline]
+    fn next(&mut self) -> Option<B> {
+        if self.flag {
+            None
+        } else {
+            let x = self.iter.next()?;
+            match (self.predicate)(x) {
+                res @ Some(_) => res,
+                None => {
+                    self.flag = true;
+                    None
+                }
+            }
+        }
```

Comment:
Can be simpler:

```rust
if self.flag {
    return None
}

let x = self.iter.next()?;
let ret = (self.predicate)(x);
self.flag = ret.is_none();
ret
```

Reply:
I think it would be even better to use `if+else` instead of `if+return`:
```rust
if self.flag {
    None
} else {
    let x = self.iter.next()?;
    let ret = (self.predicate)(x);
    self.flag = ret.is_none();
    ret
}
```